src/NITTA/LuaFrontend.hs

```diff
@@ -1,474 +1,391 @@
-{-# LANGUAGE AllowAmbiguousTypes #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE NamedFieldPuns #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE PartialTypeSignatures #-}
-{-# LANGUAGE ScopedTypeVariables #-}
-
-{-# OPTIONS -fno-warn-type-defaults #-}
-
-{- |
-Module      : NITTA.LuaFrontend
-Description : Lua frontend prototype
-Copyright   : (c) Aleksandr Penskoi, 2019
-License     : BSD3
-Maintainer  : aleksandr.penskoi@gmail.com
-Stability   : experimental
--}
+{-# OPTIONS_GHC -Wno-type-defaults #-}
+
 module NITTA.LuaFrontend (
     lua2functions,
     FrontendResult (..),
     TraceVar (..),
+
+    -- * Internal
+    AlgBuilder (..),
+    Func (..),
+    LuaValueVersion (..),
+    findStartupFunction,
+    getLuaBlockFromSources,
+    processStatement,
 ) where
 
-import Control.Monad.Identity
 import Control.Monad.State
-import Data.Bifunctor
-import qualified Data.HashMap.Strict as HM
-import Data.List (find, group, sort)
-import qualified Data.Map.Strict as M
+import qualified Data.HashMap.Strict as HashMap
+import Data.Hashable
+import qualified Data.Map.Strict as Map
 import Data.Maybe
 import Data.String
 import Data.String.ToString
-import qualified Data.String.Utils as S
-import Data.Text (Text, pack, unpack)
 import qualified Data.Text as T
 import Language.Lua
 import NITTA.Intermediate.DataFlow
 import qualified NITTA.Intermediate.Functions as F
-import NITTA.Utils (modify'_)
+import NITTA.Utils.Base
 import Text.Printf
 
+getUniqueLuaVariableName LuaValueVersion{luaValueVersionName, luaValueVersionAssignCount, luaValueVersionIsConstant} luaValueAccessCount
+    | luaValueVersionIsConstant = fromText $ "!" <> luaValueVersionName <> "#" <> showText luaValueAccessCount
+    | T.head luaValueVersionName == '_' = fromText luaValueVersionName
+    | otherwise = fromText $ luaValueVersionName <> "^" <> showText luaValueVersionAssignCount <> "#" <> showText luaValueAccessCount
+
+data Func x = Func
```

Comment:
the input and output types are different because they have different behavior. input is a particular graph edge, but output can be used in several edges.

Reply:
And about ADT type - not sure that I got your idea correctly.